File: elixir/triangle/lib/triangle.ex
defmodule Triangle do
  @type kind :: :equilateral | :isosceles | :scalene

  defguardp is_inequality(a, b, c) when a + b < c or b + c < a or a + c < b

  @doc """
  Return the kind of triangle of a triangle with 'a', 'b' and 'c' as lengths.
  """
  @spec kind(number, number, number) :: {:ok, kind} | {:error, String.t()}
  def kind(a, b, c) when a <= 0 or b <= 0 or c <= 0,
    do: {:error, "all side lengths must be positive"}

  def kind(a, b, c) when is_inequality(a, b, c),
    do: {:error, "side lengths violate triangle inequality"}

  def kind(a, a, a), do: {:ok, :equilateral}
  def kind(a, b, c) when a == b or a == c or b == c, do: {:ok, :isosceles}
  def kind(_a, _b, _c), do: {:ok, :scalene}
end

File: python/triangle/triangle.py
def equilateral(sides):
    a, b, c = sides
    return is_triangle(sides) and a == b and a == c


def isosceles(sides):
    a, b, c = sides
    return is_triangle(sides) and (a == b or a == c or b == c)


def scalene(sides):
    a, b, c = sides
    return is_triangle(sides) and (a != b and a != c and b != c)

def is_triangle(sides):
    a, b, c = sides

    return a > 0 and b > 0 and c > 0 and (a + b) >= c and (b + c) >= a and (a + c) >= b
